File: src/test/java/de/pflugradts/passbird/application/UserInterfaceAdapterPortFaker.java
package de.pflugradts.passbird.application;

import de.pflugradts.passbird.domain.model.transfer.Input;
import de.pflugradts.passbird.domain.model.transfer.Output;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.lenient;

@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class UserInterfaceAdapterPortFaker {

    private final AtomicInteger inputCount = new AtomicInteger(0);
    private final List<Input> inputList = new ArrayList<>();
    private UserInterfaceAdapterPort userInterfaceAdapterPort;

    public static UserInterfaceAdapterPortFaker faker() {
        return new UserInterfaceAdapterPortFaker();
    }

    public UserInterfaceAdapterPortFaker forInstance(final UserInterfaceAdapterPort userInterfaceAdapterPort) {
        this.userInterfaceAdapterPort = userInterfaceAdapterPort;
        return this;
    }

    public UserInterfaceAdapterPortFaker withTheseInputs(final Input... inputs) {
        inputList.addAll(Arrays.asList(inputs));
        return this;
    }

    private void givenReceivedInput() {
        lenient().when(userInterfaceAdapterPort.receive(any(Output.class))).thenAnswer(
                invocation -> inputList.get(inputCount.getAndIncrement()));
        lenient().when(userInterfaceAdapterPort.receive()).thenAnswer(
                invocation -> inputList.get(inputCount.getAndIncrement()));
    }

    public UserInterfaceAdapterPort fake() {
        givenReceivedInput();
        return userInterfaceAdapterPort;
    }

}

File: src/test/kotlin/de/pflugradts/passbird/application/UserInterfaceAdapterPortFaker.kt
package de.pflugradts.passbird.application

import de.pflugradts.passbird.domain.model.transfer.Input
import de.pflugradts.passbird.domain.model.transfer.Output
import io.mockk.every
import java.util.concurrent.atomic.AtomicInteger

fun fakeUserInterfaceAdapterPort(
    instance: UserInterfaceAdapterPort,
    withReceiveConfirmation: Boolean = false,
    withReceiveYes: Boolean = false,
    withTheseInputs: List<Input> = emptyList(),
    withTheseSecureInputs: List<Input> = emptyList(),
) {
    val inputCount = AtomicInteger(0)
    every { instance.receive() } answers { withTheseInputs[inputCount.getAndIncrement()] }
    every { instance.receive(*anyVararg<Output>()) } answers { withTheseInputs[inputCount.getAndIncrement()] }
    val secureInputCount = AtomicInteger(0)
    every { instance.receiveSecurely() } answers { withTheseSecureInputs[secureInputCount.getAndIncrement()] }
    every { instance.receiveSecurely(any<Output>()) } answers { withTheseSecureInputs[secureInputCount.getAndIncrement()] }
    every { instance.receiveConfirmation(any<Output>()) } returns withReceiveConfirmation
    every { instance.receiveYes(any<Output>()) } returns withReceiveYes
    every { instance.send(any()) } returns Unit
    every { instance.sendLineBreak() } returns Unit
    every { instance.warningSound() } returns Unit
}
